src/main/java/com/minecolonies/coremod/entity/ai/citizen/hundertaker/EntityAIWorkUndertaker.java

```diff
@@ -0,0 +1,517 @@
+     *
+     * @return the next IAIState
+     */
+    @NotNull
+    private IAIState startWorking()
+    {
+        @Nullable final BuildingGraveyard building = getOwnBuilding();
+        if (building == null || building.getBuildingLevel() < 1)
+        {
+            worker.getCitizenData().setVisibleStatus(null);
+            return IDLE;
+        }
+
+        worker.getCitizenData().setVisibleStatus(VisibleCitizenStatus.WORKING);
+        worker.getCitizenData().setIdleAtJob(false);
+
+        @Nullable final BlockPos currentGrave = building.getGraveToWorkOn();
+        if (currentGrave != null)
+        {
+            if (walkToBuilding())
+            {
+                return getState();
+            }
+
+            final TileEntity entity = world.getTileEntity(currentGrave);
+            if (entity != null && entity instanceof TileEntityGrave)
+            {
+                building.setLastGraveData((GraveData) ((TileEntityGrave) entity).getGraveData());
+                return EMPTY_GRAVE;
+            }
+            building.ClearCurrentGrave();
+        }
+
+        return WANDER;
+    }
+
+    /**
+     * The undertaker wander in the city, learning more about magic
+     *
+     * @return the next IAIState
+     */
+    @NotNull
+    private IAIState wander()
+    {
+        @Nullable final BuildingGraveyard building = getOwnBuilding();
+        if (building == null || building.getBuildingLevel() < 1)
+        {
+            worker.getCitizenData().setVisibleStatus(null);
+            return IDLE;
+        }
+
+        if(wanderPos == null)
+        {
+            final BlockPos newWanderPos = worker.getCitizenColonyHandler().getColony().getBuildingManager().
+                    getRandomBuilding(b -> b.getSchematicName() == "citizen" || b instanceof BuildingMysticalSite || b instanceof BuildingEnchanter);
+            if (newWanderPos != null)
+            {
+                wanderPos = newWanderPos;
+            }
+            return getState();
+        }
+
+        if (walkToBlock(wanderPos, 1))
+        {
+            return getState();
+        }
+
+        if(wanderPos == building.getPosition())
+        {
+            wanderPos = null;
+            worker.decreaseSaturationForContinuousAction();
+            worker.getCitizenData().getCitizenSkillHandler().addXpToSkill(getOwnBuilding().getSecondarySkill(), XP_PER_WANDER, worker.getCitizenData());
+        }
+        else
+        {
+            wanderPos = building.getPosition();
+        }
+
+        return IDLE;
+    }
+
+    private IAIState emptyGrave()
+    {
+        @Nullable final BuildingGraveyard buildingGraveyard = getOwnBuilding();
+
+        if (buildingGraveyard == null || checkForToolOrWeapon(ToolType.SHOVEL) || buildingGraveyard.getGraveToWorkOn() == null)
+        {
+            return IDLE;
+        }
+
+        worker.getCitizenData().setVisibleStatus(EMPTYING_ICON);
```

Comment:
We should probably have another walkToGrave here too, so we only get the tileEntity from the grave guaranteed when in front of it.